table/orphan_cleanup.go

```diff
@@ -0,0 +1,593 @@
+	}
+
+	dirFile, ok := f.(iceio.ReadDirFile)
+	if !ok {
+		return fmt.Errorf("directory %s does not support ReadDir", root)
+	}
+
+	entries, err := dirFile.ReadDir(-1)
+	if err != nil {
+		return err
+	}
+
+	for _, entry := range entries {
+		entryPath := filepath.Join(root, entry.Name())
+		entryInfo, err := entry.Info()
+		if err != nil {
+			continue // Skip entries we can't get info for
+		}
+		if entry.IsDir() {
+			if err := walkDirectory(fsys, entryPath, fn); err != nil {
+				return err
+			}
+		} else {
+			if err := fn(entryPath, entryInfo); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func identifyOrphanFiles(allFiles []string, referencedFiles map[string]bool, cfg *OrphanCleanupConfig) ([]string, error) {
+	var orphans []string
+
+	for _, file := range allFiles {
+		isOrphan, err := isFileOrphan(file, referencedFiles, cfg)
+		if err != nil {
+			return nil, fmt.Errorf("failed to determine if file %s is orphan: %w", file, err)
+		}
+
+		if isOrphan {
+			orphans = append(orphans, file)
+		}
+	}
+
+	return orphans, nil
+}
+
+func isFileOrphan(file string, referencedFiles map[string]bool, cfg *OrphanCleanupConfig) (bool, error) {
+	normalizedFile := normalizeFilePath(file, cfg)
+
+	if referencedFiles[file] || referencedFiles[normalizedFile] {
+		return false, nil
+	}
+
+	for referencedPath := range referencedFiles {
+		normalizedRef := normalizeFilePath(referencedPath, cfg)
+
+		if normalizedRef == normalizedFile {
+			err := checkPrefixMismatch(referencedPath, file, cfg)
+			if err != nil {
+				return false, err
+			}
+
+			return false, nil
+		}
+	}
+
+	return true, nil
+}
+
+func deleteFiles(fs iceio.IO, orphanFiles []string, cfg *OrphanCleanupConfig) ([]string, error) {
+	if len(orphanFiles) == 0 {
+		return nil, nil
+	}
+
+	if cfg.maxConcurrency == 1 {
+		return deleteFilesSequential(fs, orphanFiles, cfg)
+	}
+
+	return deleteFilesParallel(fs, orphanFiles, cfg)
+}
+
+func deleteFilesSequential(fs iceio.IO, orphanFiles []string, cfg *OrphanCleanupConfig) ([]string, error) {
+	var deletedFiles []string
+
+	for _, file := range orphanFiles {
+		var err error
+
+		if cfg.deleteFunc != nil {
+			err = cfg.deleteFunc(file)
+		} else {
+			err = fs.Remove(file)
+		}
+
+		if err != nil {
+			// Log warning but continue with other files
+			fmt.Printf("Warning: failed to delete orphan file %s: %v\n", file, err)
+
+			continue
+		}
+
+		deletedFiles = append(deletedFiles, file)
+	}
+
+	return deletedFiles, nil
+}
+
+func deleteFilesParallel(fs iceio.IO, orphanFiles []string, cfg *OrphanCleanupConfig) ([]string, error) {
+	workChan := make(chan string, len(orphanFiles))
+	resultChan := make(chan deleteResult, len(orphanFiles))
+
+	var wg sync.WaitGroup
+	numWorkers := cfg.maxConcurrency
+	if numWorkers > len(orphanFiles) {
+		numWorkers = len(orphanFiles) // Don't create more workers than files
+	}
+
+	for i := 0; i < numWorkers; i++ {
+		wg.Add(1)
+		go deleteWorker(&wg, workChan, resultChan, fs, cfg)
+	}
```

Comment:
use `errgroup` so that it fails early if there's an error and you don't need the `resultChan` at all.

Reply:
Thanks alot. It is indeed better to use errGroup